numba/core/typing/new_builtins.py

```diff
@@ -0,0 +1,1178 @@
+        signature(types.np_float32, types.np_float32, types.np_intp),
+        signature(types.np_float64, types.np_float64, types.np_intp),
+    ]
+
+
+#------------------------------------------------------------------------------
+
+
+@infer_global(bool)
+class Bool(AbstractTemplate):
+
+    def generic(self, args, kws):
+        assert not kws
+        [arg] = args
+        if isinstance(arg, (types.Boolean, types.Number)):
+            return signature(types.np_bool_, arg)
+        # XXX typing for bool cannot be polymorphic because of the
+        # types.Function thing, so we redirect to the operator.truth
+        # intrinsic.
+        return self.context.resolve_function_type(operator.truth, args, kws)
+
+
+@infer_global(int)
+class Int(AbstractTemplate):
+
+    def generic(self, args, kws):
+        if kws:
+            raise errors.NumbaAssertionError('kws not supported')
+
+        [arg] = args
+
+        if isinstance(arg, types.Integer):
+            return signature(arg, arg)
+        if isinstance(arg, (types.Float, types.Boolean)):
+            return signature(types.np_intp, arg)
+        if isinstance(arg, types.NPDatetime):
+            if arg.unit == 'ns':
+                return signature(types.np_int64, arg)
+            else:
+                raise errors.NumbaTypeError(f"Only datetime64[ns] can be converted, but got datetime64[{arg.unit}]")
+        if isinstance(arg, types.NPTimedelta):
+            return signature(types.np_int64, arg)
+
+
+@infer_global(float)
+class Float(AbstractTemplate):
+
+    def generic(self, args, kws):
+        assert not kws
+
+        [arg] = args
+
+        if arg not in types.number_domain:
+            raise errors.NumbaTypeError("float() only support for numbers")
+
+        if arg in types.np_complex_domain:
+            raise errors.NumbaTypeError("float() does not support complex")
+
+        if arg in types.integer_domain:
+            return signature(types.np_float64, arg)
+
+        elif arg in types.np_real_domain:
+            return signature(arg, arg)
+
+
+@infer_global(complex)
+class Complex(AbstractTemplate):
+
+    def generic(self, args, kws):
+        assert not kws
+
+        if len(args) == 1:
+            [arg] = args
+            if arg not in types.number_domain:
+                raise errors.NumbaTypeError("complex() only support for numbers")
+            if arg == types.np_float32:
+                return signature(types.complex64, arg)
+            else:
+                return signature(types.complex128, arg)
+
+        elif len(args) == 2:
+            [real, imag] = args
+            if (real not in types.number_domain or
+                imag not in types.number_domain):
+                raise errors.NumbaTypeError("complex() only support for numbers")
+            if real == imag == types.np_float32:
+                return signature(types.complex64, real, imag)
+            else:
+                return signature(types.complex128, real, imag)
```

Comment:
This refers to types from the "legacy" type system. Also, the `complex` constructor should return the `py_complex128` type?